src/category_theory/limits/shapes/kernels.lean

```diff
@@ -218,26 +218,37 @@ lemma kernel_not_epi_of_nonzero (w : f ≠ 0) : ¬epi (kernel.ι f) :=
 lemma kernel_not_iso_of_nonzero (w : f ≠ 0) : (is_iso (kernel.ι f)) → false :=
 λ I, kernel_not_epi_of_nonzero w $ by { resetI, apply_instance }
 
+-- TODO the remainder of this section has obvious generalizations to `has_equalizer f g`.
+
+instance has_kernel_comp_mono {X Y Z : C} (f : X ⟶ Y) [has_kernel f] (g : Y ⟶ Z) [mono g] :
+  has_kernel (f ≫ g) :=
+{ exists_limit := ⟨{
+  cone := kernel_fork.of_ι (kernel.ι f) (by simp),
+  is_limit := is_limit_aux _ (λ s, kernel.lift _ s.ι ((cancel_mono g).mp (by simp)))
+    (by tidy) (by tidy)
+}⟩ }
+
 /--
-When `g` is an isomorphism, the kernel of `f ≫ g` is isomorphic to the kernel of `f`.
+When `g` is an monomorphism, the kernel of `f ≫ g` is isomorphic to the kernel of `f`.
 -/
-def kernel_comp_is_iso {X Y Z : C} (f : X ⟶ Y) (g : Y ⟶ Z)
-  [has_kernel (f ≫ g)] [has_kernel f] [is_iso g] :
+def kernel_comp_mono {X Y Z : C} (f : X ⟶ Y) (g : Y ⟶ Z) [has_kernel f] [mono g] :
   kernel (f ≫ g) ≅ kernel f :=
 { hom := kernel.lift _ (kernel.ι _) (by { rw [←cancel_mono g], simp, }),
   inv := kernel.lift _ (kernel.ι _) (by simp), }
 
 @[simp]
-lemma kernel_comp_is_iso_hom_comp_kernel_ι {X Y Z : C} (f : X ⟶ Y) (g : Y ⟶ Z)
-  [has_kernel (f ≫ g)] [has_kernel f] [is_iso g] :
-  (kernel_comp_is_iso f g).hom ≫ kernel.ι f = kernel.ι (f ≫ g) :=
-by simp [kernel_comp_is_iso]
+lemma kernel_comp_mono_hom_comp_kernel_ι {X Y Z : C} (f : X ⟶ Y) (g : Y ⟶ Z)
+  [has_kernel f] [mono g] :
+  (kernel_comp_mono f g).hom ≫ kernel.ι f = kernel.ι (f ≫ g) :=
+by simp [kernel_comp_mono]
```

Comment:
Perhaps cleaner to mark `kernel_comp_mono` as `@[simps]`, then these lemmas shouldn't be necessary?

Reply:
Yes!